### DISCUSSION TITLE
How are "pnpMode" and "pnpFallbackMode" related ?

### DISCUSSION THREAD

**Discussion by Robin-Hoodie** [yarnpkg/berry#2462] [Q&A]
The following is my understanding from [the documentation on the yarnrc file](https://yarnpkg.com/configuration/yarnrc) on the `pnpMode` and `pnpFallbackMode` config params:

The config param `pnpMode` will determine if packages are _ever_ allowed to require dependencies that they don't list as a dependency themselves.

- In "strict" mode (the default), no packages will be allowed to require dependencies they don't list as a dependency themselves. 
- In "loose" mode, they will be allowed to require packages under some circumstances.

For "loose" mode is where the config param `pnpFallbackMode` comes into play, which will determine which kind of packages are allowed to require dependencies they don't list as a dependency themselves:

- "none": this will not allow packages to require a dependency they don't list as a dependency
- "dependencies-only": This will allow your workspace's dependencies to require a dependency that would have been present in a traditional "node_modules"-install
- "all": the same as "dependencies-only", but also applies to your workspace's code itself

I'm assuming I'm wrong in my understanding here, as a few things don't make sense:

- If `pnpFallbackMode` only applies when `pnpMode` is set to `loose`, I would find it quite strange that this does not seem to be documented explicitly anywhere
 
- If `pnpFallbackMode` set to `none` doesn't allow packages to require a dependency they don't list as a dependency themselves, that's basically the same thing as `pnpMode` set to `strict`, making this value redundant
 
- Why does `pnpFallbackMode` even have a default value, if a value of `strict` for `pnpMode` would not even make it relevant ? (I guess this one could make sense if it was documented that `pnpFallbackMode` is only relevant for `pnpMode` of `loose`)

--

Just to clarify, I am aware of the extra documentation on [PnP loose mode](https://yarnpkg.com/features/pnp#pnp-loose-mode) and [PnPFallbackMode](https://yarnpkg.com/features/pnp#different-behaviours-based-on-workspace--not-a-workspace), though that doesn't explain my points above

**Comment by arcanis** [comment]
`pnpFallbackMode` was a thing long before the loose mode, and has an effect even in strict mode (although I want to deprecate it someday). Check the last paragraph from this page for details: https://yarnpkg.com/features/pnp/#different-behaviours-based-on-workspace--not-a-workspace

**Reply by Robin-Hoodie** [reply]
Right, what I understand from there:

> if a package tries to access an unlisted dependency, we still allow to resolve it if the top-level package lists it as a dependency

I thought this only applies if `pnpMode` is set to `loose` because the documentation on `pnpMode` states the following:

> If strict (the default), Yarn won't allow modules to require packages they don't explicitly list in their own dependencies

These 2 statements seem to contradict each other

**Reply by arcanis** [reply]
It's a slight deviation. Yes, this fallback exists even in strict mode for now, but it'll eventually disappear. In the meantime, its effect is relatively small and gave us some breathing room in terms of compatibility back when we needed it (for instance, I initially implemented it because Babel wasn't able to load its plugins).

**Reply by Robin-Hoodie** [reply]
Which deviation? I'm still not really sure how `pnpFallbackMode` affects `pnpMode` to be honest

**Reply by Robin-Hoodie** [reply]
Perhaps a few specific questions that would help me understand things better:

- Does using `pnpMode=loose` and `pnpFallbackMode=none` make any sense ?
- Does using `pnpMode=strict` and `pnpFallbackMode=all` make any sense?
- For `pnpFallbackMode=dependencies-only`: what is the difference between the behaviour of `pnpMode=strict` and `pnpMode=loose`?

**Reply by arcanis** [reply]
> Does using `pnpMode=loose` and `pnpFallbackMode=none` make any sense ?

Nope

> Does using `pnpMode=strict` and `pnpFallbackMode=all` make any sense?

Yeah but ... don't. Basically you would only do this if you want to break the encapsulation a little, but not entirely. Weird.

> For `pnpFallbackMode=dependencies-only`: what is the difference between the behaviour of `pnpMode=strict` and `pnpMode=loose`

The set of packages that your *dependencies* can read by bypassing the hoisting. Even in loose mode your application (ie your own code, that you wrote yourself) isn't allowed to access unlisted dependencies. Only with `pnpFallbackMode=all` will your application be allowed to access dependencies through hoisting.

**Reply by Robin-Hoodie** [reply]
Right, thanks for clarifying the first 2 examples, though I'm honestly still fuzzy on some details

Let's say I have a dependency on package "A" and package "B".
Package "A" has a dependency on package "C"

The following piece of code is part of my application code/workspace:

`require('C")`

Yarn will refuse to require "C" for the following settings

- `pnpMode=strict`, `pnpFallbackMode` will not make this work, even if it's set to `all`
- `pnpMode=loose`, `pnpFallbackMode=none|dependencies-only`

This will only work by setting `pnpMode=loose` and `pnpFallbackMode=all`

The following piece of code is part of package "A":

`require("B")`

Yarn will refuse to require "B" if `pnpFallbackMode` is set to `none`

It will work if `pnpFallbackMode` is set to "dependencies-only" or "all", regardless of what `pnpMode` is set to

Is that correct?

**Comment by merceyz** [comment]
To put it differently `pnpMode` controls what is in the fallback pool, `pnpFallbackMode` controls what is allowed to access the fallback pool.
